File: frontend/src/lib/diarizationSettings.ts
/**
 * Speaker-diarization UI settings (task D5).
 *
 * Persisted client-side in localStorage (single JSON blob, mirroring the
 * betaFeatures pattern) and pushed into the Rust backend via
 * `set_diarization_settings` so the recording / auto-diarize paths honor them.
 * Defaults intentionally match the backend's env/compiled defaults, so an
 * un-synced first run behaves identically to D2/D3.
 */

import { invoke } from '@tauri-apps/api/core';

export interface DiarizationSettings {
  /** Master switch for speaker diarization. */
  enabled: boolean;
  /** Run diarization automatically when a meeting is saved. */
  autoDiarize: boolean;
  /** Persist separate mic/system tracks (D2) for better diarization. */
  saveSeparateTracks: boolean;
  /** Prefix each line with the speaker name in the transcript sent to the LLM. */
  summarizeWithSpeakers: boolean;
}

export const DEFAULT_DIARIZATION_SETTINGS: DiarizationSettings = {
  enabled: true,
  autoDiarize: true,
  saveSeparateTracks: true,
  summarizeWithSpeakers: false,
};

const STORAGE_KEY = 'diarizationSettings';

/** Load settings from localStorage, falling back to defaults. SSR-safe. */
export function loadDiarizationSettings(): DiarizationSettings {
  if (typeof window === 'undefined') return { ...DEFAULT_DIARIZATION_SETTINGS };
  try {
    const raw = window.localStorage.getItem(STORAGE_KEY);
    if (!raw) return { ...DEFAULT_DIARIZATION_SETTINGS };
    const parsed = JSON.parse(raw);
    return { ...DEFAULT_DIARIZATION_SETTINGS, ...parsed };
  } catch {
    return { ...DEFAULT_DIARIZATION_SETTINGS };
  }
}

/** Persist settings to localStorage. */
export function saveDiarizationSettings(settings: DiarizationSettings): void {
  if (typeof window === 'undefined') return;
  try {
    window.localStorage.setItem(STORAGE_KEY, JSON.stringify(settings));
  } catch (err) {
    console.warn('Failed to persist diarization settings:', err);
  }
}

/** Push the three backend-facing toggles into the Rust runtime. */
export async function syncDiarizationSettingsToBackend(
  settings: DiarizationSettings = loadDiarizationSettings()
): Promise<void> {
  try {
    await invoke('set_diarization_settings', {
      enabled: settings.enabled,
      autoDiarize: settings.autoDiarize,
      saveSeparateTracks: settings.saveSeparateTracks,
    });
  } catch (err) {
    // Non-fatal: backend keeps its env/default values.
    console.warn('Failed to sync diarization settings to backend:', err);
  }
}

File: frontend/src-tauri/src/audio/diarization_settings.rs
//! Runtime toggles for diarization features (task D5).
//!
//! Backs three user-facing settings that the frontend persists (tauri-plugin
//! store / localStorage) and pushes into the backend via
//! [`set_diarization_settings`]:
//!   - `enabled`              — master switch for diarization.
//!   - `auto_diarize`         — run diarization automatically when a meeting is
//!     saved.
//!   - `save_separate_tracks` — persist `mic.mp4` + `system.mp4` (D2) for higher
//!     diarization quality.
//!
//! Each value is a tri-state atomic: `UNSET` means "fall back to the environment
//! variable / compiled default" (so behavior is unchanged for D2/D3 users who
//! never touch the UI), otherwise the explicit boolean the UI last pushed. This
//! keeps the change additive: nothing in the recording / diarization hot paths
//! needs a DB pool or an `AppHandle`, they just read a process-global atomic.

use serde::{Deserialize, Serialize};
use std::sync::atomic::{AtomicI8, Ordering};

/// Tri-state sentinel: the UI has not set an explicit value yet.
const UNSET: i8 = -1;

static ENABLED: AtomicI8 = AtomicI8::new(UNSET);
static AUTO_DIARIZE: AtomicI8 = AtomicI8::new(UNSET);
static SAVE_SEPARATE_TRACKS: AtomicI8 = AtomicI8::new(UNSET);

/// Read a tri-state cell, deferring to `fallback` while it is `UNSET`.
fn resolve(cell: &AtomicI8, fallback: impl FnOnce() -> bool) -> bool {
    match cell.load(Ordering::Relaxed) {
        0 => false,
        1 => true,
        _ => fallback(),
    }
}

/// Write an explicit boolean into a tri-state cell.
fn store(cell: &AtomicI8, value: bool) {
    cell.store(i8::from(value), Ordering::Relaxed);
}

/// Master diarization switch. Defaults to enabled until the UI says otherwise.
pub fn enabled() -> bool {
    resolve(&ENABLED, || true)
}

/// Whether to auto-diarize on meeting save. Falls back to the
/// `MEETILY_AUTO_DIARIZE` environment default (task D3).
pub fn auto_diarize() -> bool {
    resolve(&AUTO_DIARIZE, crate::audio::diarization::auto_diarize_enabled)
}

/// Whether to persist separate mic/system tracks. Falls back to the
/// `MEETILY_SAVE_SEPARATE_TRACKS` environment default (task D2).
pub fn save_separate_tracks() -> bool {
    resolve(
        &SAVE_SEPARATE_TRACKS,
        crate::audio::track_saver::env_save_separate_tracks,
    )
}

/// Serializable view of the three toggles (as currently resolved).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiarizationSettings {
    pub enabled: bool,
    pub auto_diarize: bool,
    pub save_separate_tracks: bool,
}

fn snapshot() -> DiarizationSettings {
    DiarizationSettings {
        enabled: enabled(),
        auto_diarize: auto_diarize(),
        save_separate_tracks: save_separate_tracks(),
    }
}

/// Return the effective diarization toggles (UI override or env/default).
#[tauri::command]
pub async fn get_diarization_settings() -> Result<DiarizationSettings, String> {
    Ok(snapshot())
}

/// Push the UI's toggle values into the backend runtime. Additive: recording
/// (`save_separate_tracks`) and auto-diarize honor these immediately without a
/// restart. Persistence across restarts lives in the frontend store; the
/// frontend re-pushes on startup.
#[tauri::command]
pub async fn set_diarization_settings(
    enabled: bool,
    auto_diarize: bool,
    save_separate_tracks: bool,
) -> Result<DiarizationSettings, String> {
    store(&ENABLED, enabled);
    store(&AUTO_DIARIZE, auto_diarize);
    store(&SAVE_SEPARATE_TRACKS, save_separate_tracks);
    Ok(snapshot())
}

#[cfg(test)]
mod tests {
    // Exercises the tri-state logic on a local cell to avoid touching the
    // process-global statics (which other modules' tests also read).
    use super::{resolve, store, UNSET};
    use std::sync::atomic::AtomicI8;

    #[test]
    fn unset_defers_to_fallback() {
        let cell = AtomicI8::new(UNSET);
        assert!(resolve(&cell, || true));
        assert!(!resolve(&cell, || false));
    }

    #[test]
    fn explicit_value_wins_over_fallback() {
        let cell = AtomicI8::new(UNSET);
        store(&cell, false);
        assert!(!resolve(&cell, || true));
        store(&cell, true);
        assert!(resolve(&cell, || false));
    }
}
